src/main/java/uk/ac/cam/cl/dtg/segue/dao/kafkaStreams/UserStatisticsStreamsApplication.java

```diff
@@ -0,0 +1,535 @@
+                            TimeUnit.MILLISECONDS));
+
+
+                } else if (daysSinceStreakIncrease == 1) {
+
+                    Integer curActivity = snapshotRecord.path("streak_record").path("current_activity").asInt();
+
+                    if (curActivity != snapshotRecord.path("streak_record").path("activity_threshold").asInt()) {
+                        streakRecord.put("currentActivity", snapshotRecord.path("streak_record").path("current_activity").asInt());
+                    }
+
+                    streakRecord.put("currentStreak", TimeUnit.DAYS.convert(streakEndTimestamp - snapshotRecord.path("streak_record").path("streak_start").asLong(),
+                            TimeUnit.MILLISECONDS));
+
+                    streakRecord.put("dailyStreakMessage", "Complete your daily question goal by the end of today to increase your streak!");
+                }
+
+                streakRecord.put("largestStreak", snapshotRecord.path("streak_record").path("largest_streak").asLong());
+                userSnapshot.put("streakRecord", streakRecord);
+            }
+
+            // for all users who aren't students, augment the snapshot with additional teacher stats
+            if (!user.getRole().equals(Role.STUDENT) && snapshotRecord.has("teacher_record")) {
+                userSnapshot.put("teacherActivityRecord", snapshotRecord.path("teacher_record"));
+            }
+        }
+
+        return userSnapshot;
+    }
+
+
+
+    /**
+     *
+     * @param activityType
+     * @param teacherRecord
+     * @return
+     */
+    private JsonNode updateTeacherActivityRecord(String activityType, JsonNode teacherRecord) {
+
+        Integer count = teacherRecord.path(activityType).asInt();
+        ((ObjectNode) teacherRecord).put(activityType, count + 1);
+
+        return teacherRecord;
+    }
+
+
+
+
+
+
+    /**
+     *We call this method to update the streak data for the user snapshot record
+     *
+     * @param userId id of the user we want to update
+     * @param latestEvent json object describing the event which triggers the streak update
+     * @param streakRecord the current snapshot of the streak record
+     * @return the new updated streak record
+     */
+    private JsonNode updateStreakRecord(String userId, JsonNode latestEvent, JsonNode streakRecord) {
+
+        // timestamp of streak start
+        Long streakStartTimestamp = streakRecord.path("streak_start").asLong();
+
+        // timestamp of streak end
+        Long streakEndTimestamp = streakRecord.path("streak_end").asLong();
+
+        // timestamp of latest event
+        Long latestEventTimestamp = latestEvent.path("timestamp").asLong();
+
+        // set up calendar objects for date comparisons
+        Calendar latest = Calendar.getInstance();
+        latest.setTimeInMillis(latestEventTimestamp);
+        latest = roundDownToDay(latest);
+
+
+        // 1) If the current activity threshold for the day has been reached
```

Comment:
These comments break up the logic of the code well and so, in future, would be good candidates for where to break the method down and split into smaller subroutines/methods.